src/app/clusters/thread-border-router-management-server/thread-border-router-management-server.cpp

```diff
@@ -0,0 +1,384 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "thread-border-router-management-server.h"
+
+#include "app-common/zap-generated/cluster-objects.h"
+#include "app-common/zap-generated/ids/Attributes.h"
+#include "app-common/zap-generated/ids/Clusters.h"
+#include "app-common/zap-generated/ids/Commands.h"
+#include "app/AttributeAccessInterfaceRegistry.h"
+#include "app/AttributeValueEncoder.h"
+#include "app/CommandHandler.h"
+#include "app/CommandHandlerInterface.h"
+#include "app/InteractionModelEngine.h"
+#include "app/MessageDef/StatusIB.h"
+#include "app/clusters/general-commissioning-server/general-commissioning-server.h"
+#include "app/data-model/Nullable.h"
+#include "crypto/RandUtils.h"
+#include "lib/core/CHIPError.h"
+#include "lib/support/CodeUtils.h"
+#include "lib/support/Span.h"
+#include "lib/support/ThreadOperationalDataset.h"
+#include "platform/CHIPDeviceEvent.h"
+#include "platform/PlatformManager.h"
+#include "protocols/interaction_model/StatusCode.h"
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ThreadBorderRouterManagement {
+
+using Protocols::InteractionModel::Status;
+
+static bool IsOverCASESession(CommandHandlerInterface::HandlerContext & ctx)
+{
+    Messaging::ExchangeContext * exchangeCtx = ctx.mCommandHandler.GetExchangeContext();
+    return exchangeCtx && exchangeCtx->HasSessionHandle() && exchangeCtx->GetSessionHandle()->IsSecureSession() &&
+        exchangeCtx->GetSessionHandle()->AsSecureSession()->GetSecureSessionType() == Transport::SecureSession::Type::kCASE;
+}
+
+Status ServerInstance::HandleGetDatasetRequest(bool isOverCASESession, Delegate::DatasetType type,
+                                               Thread::OperationalDataset & dataset)
+{
+    if (!isOverCASESession)
+    {
+        return Status::UnsupportedAccess;
+    }
+
+    CHIP_ERROR err = mDelegate->GetDataset(dataset, type);
+    if (err != CHIP_NO_ERROR)
+    {
+        return err == CHIP_IM_GLOBAL_STATUS(NotFound) ? StatusIB(err).mStatus : Status::Failure;
+    }
+    return Status::Success;
+}
+
+Status ServerInstance::HandleSetActiveDatasetRequest(bool failSafeArmed,
+                                                     const Commands::SetActiveDatasetRequest::DecodableType & req)
+{
+    // The SetActiveDatasetRequest command SHALL be FailSafeArmed. Upon receiving this command, the Thread BR will set its
+    // active dataset. If the dataset is set successfully, OnActivateDatasetComplete will be called with CHIP_NO_ERROR, prompting
+    // the Thread BR to respond with a success status and disarm the FailSafe timer. If an error occurs while setting the active
```

Comment:
> Thread BR to respond with a success status and disarm the FailSafe timer
 
It's always the client that disarms fail safe, and it can only be done via `CommissioningComplete`. I don't think I understand the comment.

Reply:
Add a TODO of the SPEC issue https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/10022